### ISSUE TITLE
Bug with GNU make on windows jobs

### ISSUE THREAD

**Issue by felipec** [actions/runner-images#7253]
### Description

It took me a long time to figure out what was the problem, but it seems there's a problem in `make` when the configured shell ends in `.exe`, for example `bash` works, but not `bash.exe`, and the shell I see by default is `sh.exe`.

I tried configuring the sell manually:

    make SHELL='C:/Program Files/Git/usr/bin/bash'

And that works, but only on the top level Makefile, when doing a submake with `$(MAKE) -C` the variable is reset to `sh.exe`, and then I see this error:

```
Windows Subsystem for Linux has no installed distributions.

Distributions can be installed by visiting the Microsoft Store:

https://aka.ms/wslstore
```

So presumably `make` is ignoring the default `SHELL` variable and running `C:/Windows/System32/bash.exe` for some reason.

I found this curious message online for CMake:

```
  sh.exe was found in your PATH, here:

  C:/Program Files/Git/usr/bin/sh.exe

  For MinGW make to work correctly sh.exe must NOT be in your path.

  Run cmake from a shell that does not have sh.exe in your PATH.

  If you want to use a UNIX shell, then use MSYS Makefiles.
```

I see a ton of people mentioning, but no one explains **why** make doesn't work, but I can confirm that it doesn't.

### Platforms affected

- [ ] Azure DevOps
- [X] GitHub Actions - Standard Runners
- [ ] GitHub Actions - Larger Runners

### Runner images affected

- [ ] Ubuntu 18.04
- [ ] Ubuntu 20.04
- [ ] Ubuntu 22.04
- [ ] macOS 11
- [ ] macOS 12
- [X] Windows Server 2019
- [X] Windows Server 2022

### Image version and build link

2.302.1

### Is it regression?

I don't know

### Expected behavior

I would expect `make` to just work (even with sub-makes).

### Actual behavior

`make` doesn't work, it's using a wrong shell (not bash).

### Repro steps

1. Run a job with windows-latest
2. Run make

If you need an way to reproduce this, I probably can create a very simple repo that reproduces it, but any project using `make` should be affected.

**Comment by felipec** [comment]
I forgot to mention something else, the failure only happens when the Makefile runs a bash script directly, like:

    ./script.sh

If it's a command it works fine (IIRC), like:

    bash script.sh

Or:

    ./script.sh || true

**Comment by sergei-pyshnoi** [comment]
Hello @felipec. Can you please provide more detailed repro steps with code and pipeline example?

**Comment by felipec** [comment]
@sergei-pyshnoi OK, I created a very simple repo that reproduces the issue: [felipec/make-simple](https://github.com/felipec/make-simple). It literally does nothing but call a simple script in many ways.

Here's a [test run](https://github.com/felipec/make-simple/actions/runs/4363674891/jobs/7630055426). As you can see it works perfectly fine in ubuntu, but not windows.

```
## info
main SHELL C:/Program Files/Git/bin/sh.exe
main MAKE C:/ProgramData/Chocolatey/lib/mingw/tools/install/mingw64/bin/mingw32-make.exe
## test
sub/foo.t
Windows Subsystem for Linux has no installed distributions.

mingw32-make: *** [makefile:[10]: test] Error 1
Distributions can be installed by visiting the Microsoft Store:

https://aka.ms/wslstore

Error: Process completed with exit code 1.
```

The only way to make the test in the subdir work is to specify `-C sub` in the command, and the `SHELL` variable:

    make SHELL='C:/Program Files/Git/usr/bin/bash' -C sub

Here's [a run](https://github.com/felipec/make-simple/actions/runs/4363782580/jobs/7630293579) where that works.

I don't know how to tell GitHub actions to run all the steps regardless (I'm don't know powershell, although I guess I could use bash and do `|| true`).

**Comment by felipec** [comment]
Ah, `make -i` ignores all errors and now I was able to test all the steps in [this run](https://github.com/felipec/make-simple/actions/runs/4363884985/jobs/7630527960).

| test                | bin/sh.exe | usr/bin/bash |
| -------------                       | ---- |----|
| test (`sub/foo.t`)                    | Fail | OK |
| straight (`bash -c sub/foo.t`)        | OK   | OK |
| sub-test (`./foo.t`)                  | Fail | OK |
| sub-straight-test (`bash -c ./foo.t`) | OK   | OK |

As you can see the default shell just doesn't work. I don't know how `make` is picking it.

In theory I could just do `make SHELL='C:/Program Files/Git/usr/bin/bash'` but as you can see in the step "Test with shell" that variable doesn't get carried to the sub-test (`$(MAKE) -C sub test`), I don't know why.

**Comment by sergei-pyshnoi** [comment]
Thanks for detailed information. We will take a look.

**Comment by ilia-shipitsin** [comment]
@felipec , git bash is "mingw" app while "make" is native win app. I'm afraid that here we observe one of mingw interoperation side effects ((

**Comment by felipec** [comment]
What do you mean native win app? I did `make -v` everywhere (on the runner shell, in the make shell, and even `$(MAKE) -v`, **all** of them show the same version:

```
GNU Make 4.3
Built for x86_64-w64-mingw32
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by felipec** [comment]
As you can see from [this job](https://github.com/felipec/make-simple/actions/runs/4384166490/jobs/7675306368), the issue doesn't just happen with `sh.exe`, but `bash.exe` as well.

This works: `'C:/Program Files/Git/bin/sh'`, this doesn't `'C:/Program Files/Git/bin/sh.exe'`.

Maybe `make` is checking the extension and doing something different?

**Comment by ilia-shipitsin** [comment]
I planned to run make with debug `-d` and compare output. Maybe it will give some clue

**Comment by felipec** [comment]
@ilia-shipitsin Here it is: [test with debug](https://github.com/felipec/make-simple/actions/runs/4384434654/jobs/7675905588).

The interesting part is this with `sh`:

```
CreateProcess(NULL,C:/Program Files/Git/bin/sh -c sub/foo.t,...)
```

And this with `sh.exe`:

```
CreateProcess(NULL,bash D:\a\make-simple\make-simple\sub\foo.t,...)
```

**Comment by ilia-shipitsin** [comment]
can you try `bash` as well in order to compare "bash" and "bash" ?

**Comment by felipec** [comment]
I'm not sure I follow, but I called `bash sub/foo.t` directly in [this job](https://github.com/felipec/make-simple/actions/runs/4387547754/jobs/7682973880).

With `C:/Program Files/Git/bin/sh` it's:

    CreateProcess(NULL,C:/Program Files/Git/bin/sh -c "bash sub/foo.t",...)

With `C:/Program Files/Git/bin/sh.exe` it's:

    CreateProcess(C:\Program Files\Git\bin\bash.exe,bash sub/foo.t,...)

**Comment by sergei-pyshnoi** [comment]
Hello @felipec . Sorry for the late response. Please, can you take a look on [result of this job](https://github.com/sergei-pyshnoi/make-simple/actions/runs/4448018483/jobs/7810262060) . Look like every tests was passed. I am just changed `#!/bin/bash` => `#!/bin/sh` in the script. Look like it was reason why is `make` choses wrong executor of shell.

**Comment by felipec** [comment]
> Look like every tests was passed. I am just changed #!/bin/bash => #!/bin/sh in the script.

Yes, I thought that might work, but `/bin/bash` should work as well.

> Look like it was reason why is `make` choses wrong executor of shell.

No, this comes from the Unix world, a script that starts with `#!/bin/bash` is supposed to be interpreted by a program `/bin/bash`: see [shebang](https://en.wikipedia.org/wiki/Shebang_(Unix)).

For example a program `/tmp/test.perl` like this:

```
#!/bin/perl

print("foo\n")
```

Is suppose to be interpreted by perl, so doing `/tmp/test.perl` or `perl /tmp/test.perl` should be equivalent.

If the shebang was `#!/bin/python` then that should be the interpreter, but it could even be `#!/bin/cat`.

GNU make isn't making a mistake here, a `foo.t` script that starts `!#/bin/bash` is meant to be interpreted by bash. Note that in some Linux distributions this does make a difference, as `/bin/sh` could be a POSIX-compatible shell, not necessarily bash, so a script with bash-specific features would fail with `sh` but succeed with `bash`.

---

The question is why does this fail:

    CreateProcess(NULL,bash D:\a\make-simple\make-simple\sub\foo.t,...)

But this succeed:

    CreateProcess(NULL,sh D:\a\make-simple\make-simple\sub\bar.t,...)

**Comment by sergei-pyshnoi** [comment]
@felipec After some testing, I found that `GNU make` + `GitBash` works a little differently on Windows than it does on UNIX systems. For using `bash` as interpreter in this case you can put `#!/usr/bin/env bash` as shebang to your script. Here is [success result](https://github.com/sergei-pyshnoi/make-simple/actions/runs/4469318174/jobs/7851252666) of running `make` command.

**Comment by work-direxfranko** [comment]
Hsx

**Comment by work-direxfranko** [comment]
Seguridad

**Comment by sergei-pyshnoi** [comment]
@felipec . After investigation, we found that is after disabling  WSL feature, `make` starts running the correct bash. Other setting like installation path end PATH variable is correct.  The main problem is how exactly `make` running bash process. Unfortunately we can not do anything with this and WSL can`t be disabled in runtime. Please use provided workarounds. I am closing issue for now, but if you have additional question feel free to ask.

**Comment by felipec** [comment]
I've managed to narrow down the problem, this reproduces the issue:

```
$env:Path = "C:/Windows/System32;$env:Path"
bash --version
```

No need to run `make` at all.

In this case the `PATH` environment variable can be fiddled to make it work, but when using [CreateProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa) that's not an option because it prioritizes the 32-bit system directory regardless of `PATH`. And that's what `make` uses.

So the options are:

1. Fix WSL: microsoft/WSL#13277
1. Remove `C:/Windows/System32/bash.exe`
1. Disable WSL: #12646
